searchpath: add pg_extension to searchpath.Iter() before public

otan opened the pull request:
Release note (sql change): When doing name resolution via search path,
the `pg_extension` schema (containing tables such as `geometry_columns`,
`geography_columns` and `spatial_ref_sys`) will have an attempted
resolution before the `public` schema. This mimics PostGIS behavior
where the aforementioned tables are in the public schema, and so
by default are discoverable tables with a new CLI session.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49611)
<!-- Reviewable:end -->

rohany commented on pkg/sql/sqlbase/constants.go:
```diff
@@ -17,7 +17,7 @@ import (
 
 // DefaultSearchPath is the search path used by virgin sessions.
-var DefaultSearchPath = sessiondata.MakeSearchPath([]string{"public"})
+var DefaultSearchPath = sessiondata.MakeSearchPath([]string{"public", "pg_extension"})
```
Why do we include pg_extension here? I thought the whole point was to have pg_extension implicitly resolve before public?

otan replied:
accident!

rohany commented on pkg/sql/sessiondata/search_path.go:
```diff
@@ -211,6 +225,13 @@ func (iter *SearchPathIter) Next() (path string, ok bool) {
 			return iter.tempSchemaName, true
 		}
+		// pg_extension should be read before delving into the schema.
+		if iter.paths[iter.i-1] == "public" && iter.implicitPgExtension {
```
Use the constant `PublicSchemaName`?

otan replied:
can't - circular dependency.

otan replied:
well i can move stuff around....update incoming!

rohany commented:
LGTM, but is there also a logic test you can include that accesses these tables the way that you want?

otan commented:
thought i had one, but i stashed it accidentally. good catch.

otan committed:
searchpath: add pg_extension to searchpath.Iter() before public

Release note (sql change): When doing name resolution via search path,
the `pg_extension` schema (containing tables such as `geometry_columns`,
`geography_columns` and `spatial_ref_sys`) will have an attempted
resolution before the `public` schema. This mimics PostGIS behavior
where the aforementioned tables are in the public schema, and so
by default are discoverable tables with a new CLI session.

rohany commented:
LGTM

otan commented:
tftr!

bors r=rohany
